module/zfs/vdev.c

```diff
@@ -1616,53 +1705,85 @@ vdev_uses_zvols(vdev_t *vd)
 	return (B_FALSE);
 }
 
-void
-vdev_open_children(vdev_t *vd)
+/*
+ * Returns B_TRUE if the passed child should be opened.
+ */
+static boolean_t
+vdev_default_open_children_func(vdev_t *vd)
+{
+	return (B_TRUE);
+}
+
+/*
+ * Open the requested child vdevs.  If any of the leaf vdevs are using
+ * a ZFS volume then do the opens in a single thread.  This avoids a
+ * deadlock when the current thread is holding the spa_namespace_lock.
+ */
+static void
+vdev_open_children_impl(vdev_t *vd, vdev_open_children_func_t *open_func)
 {
-	taskq_t *tq;
 	int children = vd->vdev_children;
 
-	/*
-	 * in order to handle pools on top of zvols, do the opens
-	 * in a single thread so that the same thread holds the
-	 * spa_namespace_lock
-	 */
-	if (vdev_uses_zvols(vd)) {
-retry_sync:
-		for (int c = 0; c < children; c++)
-			vd->vdev_child[c]->vdev_open_error =
-			    vdev_open(vd->vdev_child[c]);
-	} else {
-		tq = taskq_create("vdev_open", children, minclsyspri,
-		    children, children, TASKQ_PREPOPULATE);
-		if (tq == NULL)
-			goto retry_sync;
+	taskq_t *tq = taskq_create("vdev_open", children, minclsyspri,
+	    children, children, TASKQ_PREPOPULATE);
+	vd->vdev_nonrot = B_TRUE;
 
-		for (int c = 0; c < children; c++)
+	for (int c = 0; c < children; c++) {
+		vdev_t *cvd = vd->vdev_child[c];
+
+		if (open_func(cvd) == B_FALSE)
+			continue;
+
+		if (tq == NULL || vdev_uses_zvols(vd)) {
+			cvd->vdev_open_error = vdev_open(cvd);
+		} else {
 			VERIFY(taskq_dispatch(tq, vdev_open_child,
-			    vd->vdev_child[c], TQ_SLEEP) != TASKQID_INVALID);
+			    cvd, TQ_SLEEP) != TASKQID_INVALID);
+		}
+
+		vd->vdev_nonrot &= cvd->vdev_nonrot;
+	}
 
+	if (tq != NULL) {
+		taskq_wait(tq);
 		taskq_destroy(tq);
 	}
+}
 
-	vd->vdev_nonrot = B_TRUE;
+/*
+ * Open all child vdevs.
+ */
+void
+vdev_open_children(vdev_t *vd)
+{
+	vdev_open_children_impl(vd, vdev_default_open_children_func);
+}
 
-	for (int c = 0; c < children; c++)
-		vd->vdev_nonrot &= vd->vdev_child[c]->vdev_nonrot;
+/*
+ * Conditionally open a subset of child vdevs.
+ */
+void
+vdev_open_children_subset(vdev_t *vd, vdev_open_children_func_t *open_func)
+{
+	vdev_open_children_impl(vd, open_func);
 }
 
 /*
- * Compute the raidz-deflation ratio.  Note, we hard-code
- * in 128k (1 << 17) because it is the "typical" blocksize.
- * Even though SPA_MAXBLOCKSIZE changed, this algorithm can not change,
- * otherwise it would inconsistently account for existing bp's.
+ * Compute the raidz-deflation ratio.  Note, we used to hard-code
+ * in 128k (1 << 17) because it is/was the "typical" blocksize.
+ * The "typical" blocksize can now be set using zfs_deflate_shift.
+ * Once defined for a new pool, the value cannot be changed,
+ * otherwise we would inconsistently account for existing bp's.
```

Comment:
Although each top-level vdev does have a vdev_deflate_ratio field, they are all set to the same value by vdev_set_deflate_ratio(). In the past, this was a fixed value. With this change, it is computed from spa_deflate.

Reply:
I think you're saying that every top-level vdev's `vdev_deflate_ratio` is the same. So if you create a pool with one RAIDZ device, and then add a mirror device, the mirror will have the same deflate ratio as the RAIDZ device, which would be incorrect.  Which code implements that?  I'm reading the new version of `vdev_set_deflate_ratio()` but it looks like it's calculating a different ratio for each vdev.

```
		vd->vdev_deflate_ratio = typical /
 		    (vdev_psize_to_asize(vd, typical) >> SPA_MINBLOCKSHIFT);
```

Was vdev_psize_to_asize() changed to return the same value for every vdev in the pool?

> The change "co-ops" the value stored in DMU_POOL_DEFLATE. We now store not just 0 or 1, but an actual deflation ratio here.

The code in `vdev_set_deflate_ratio()` seems to be inconsistent with this description. It's using `spa_deflate` as though it contained the shift of the typical block size, not the ratio.
```
			typical = 1 << vd->vdev_spa->spa_deflate;
```

Reply:
Heh, I suppose I should have been clearer in my explanation. The value computed by vdev_set_deflate_ratio() is computed in the same way as was done previously. The resulting value will be the same for all top-level vdevs with the same psize_to_asize. vdev_psize_to_asize() was not modified to return the same value for every vdev in the pool (that would be silly).

You are correct, the value actually stored in spa_deflate is the shift of the typical block size, not the actual deflate ratio.

Reply:
OK that makes sense.  In that case I think the only issue is backwards compatibility of the on-disk format.

Reply:
Since the `spa_deflate` changes are really orthogonal to dRAID I've gone ahead and pulled them out of the PR.  They're still something I think we want to do but we can tackle it in a dedicated PR and work through the on-disk format concerns there.